commit 2d856dbb0cb07176cc1b9d8be6e4f7e2686ad4e1 (refs/changes/43/5143/3)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-11-29T17:40:32+01:00 (10 months ago)
    
    OS-7395 bhyve: VM_{SUSPEND,RESUME}_CPU ioctls ignore vcpu argument

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index e6212a84ea..0960337923 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -969,10 +969,30 @@ vmmdev_do_ioctl(vmm_softc_t *sc, int cmd, intptr_t arg, int md,
 	case VM_ACTIVATE_CPU:
 		error = vm_activate_cpu(sc->vmm_vm, vcpu);
 		break;
+
 	case VM_SUSPEND_CPU:
+		if (ddi_copyin(datap, &vcpu, sizeof (vcpu), md)) {
+			error = EFAULT;
+			break;
+		}
+		if (vcpu < -1 || vcpu >= VM_MAXCPU) {
+			error = EINVAL;
+			break;
+		}
+
 		error = vm_suspend_cpu(sc->vmm_vm, vcpu);
 		break;
+
 	case VM_RESUME_CPU:
+		if (ddi_copyin(datap, &vcpu, sizeof (vcpu), md)) {
+			error = EFAULT;
+			break;
+		}
+		if (vcpu < -1 || vcpu >= VM_MAXCPU) {
+			error = EINVAL;
+			break;
+		}
+
 		error = vm_resume_cpu(sc->vmm_vm, vcpu);
 		break;
 
